### PR TITLE
Move ability to save pool allocation data to json into the RHI

### PR DESCRIPTION
**Pull request by rgba16f**
Enable the RHI to dump a data file containing the current state of gpu memory allocations in response to asserts & other events.

### CODE REVIEW

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/MemoryStatistics.h [hunk 1]
```diff
@@ -13,85 +13,130 @@
...
+
+    //! Utility function to write captured pool data to a json document
+    void WritePoolsToJson(AZStd::vector<MemoryStatistics::Pool>& pools, rapidjson::Document& doc, rapidjson::Value& docRoot);
+
+    //! Utility function to trigger an emergency dump of pool information to json.
+    //! intended to be used for gpu memory failure debugging.
+    void DumpPoolInfoToJson();
+
+#ifndef AZ_RELEASE_BUILD
+    #define AZ_RHI_DUMP_POOL_INFO_ON_FAIL(test) \
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/MemoryStatistics.h:131]: Minor : Rename test to result.

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        }
+        docRoot.AddMember(PoolsAttribStr, poolsArray, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionMajorAttribStr, 1, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionMinorAttribStr, 0, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionRevisionAttribStr, 0, doc.GetAllocator());
+    }
+
+    void DumpPoolInfoToJson()
+    {
+            AZ::IO::Path path = AZStd::string_view(AZ::Utils::GetO3deLogsDirectory());
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp:110]: We probably dont want to always dump on a failed allocation by default. Not sure what happens if this code runs on android/ios. Maybe hide this behind a cvar? So if a developer is trying to debug a gpu OOM crash he can enable the cvar and re-run the app.

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+            WritePoolsToJson(pools, doc, root);
+
+            rapidjson::StringBuffer jsonStringBuffer;
+            rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(jsonStringBuffer);
+            doc.Accept(writer);
+
+            outputFile.Write(jsonStringBuffer.GetString(), jsonStringBuffer.GetSize());
+            outputFile.Close();
+
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp:157]: Minor: Remove extra line

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            time_t ltime;
+            time(&ltime);
+            tm today;
+#if AZ_TRAIT_USE_SECURE_CRT_FUNCTIONS
+            localtime_s(&today, &ltime);
+#else
+            today = *localtime(&ltime);
+#endif
+            char sTemp[128];
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp:125]: Minor: Rename sTemp to gpuMemDumpFileName

**Review by moudgils — COMMENTED** [review]
What are your thoughts on moving DumpPoolInfoToJson to an ebus event that can be called by anyone instead of calling it directly from imguigpu profiler? I can see use cases where by any gem may want to do add custom code to dump out gpu mem information via code as part of debugging gpu mem related issues.

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+#include <AzCore/JSON/stringbuffer.h>
+#include <AzCore/JSON/prettywriter.h>
+#include <AzCore/Utils/Utils.h>
+
+#include <Atom/RHI/RHISystemInterface.h>
+
+namespace AZ::RHI
+{
+    // POOL attributes
```
Comment (gadams3) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp:18]:
```suggestion
    // Pool attributes
```

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                poolObject.AddMember(MemoryTypeAttribStr, HostMemoryTypeValueStr, doc.GetAllocator());
+                poolObject.AddMember(BudgetInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageHost.m_budgetInBytes), doc.GetAllocator());
+                poolObject.AddMember(TotalResidentInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageHost.m_totalResidentInBytes.load()), doc.GetAllocator());
+                poolObject.AddMember(UsedResidentInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageHost.m_usedResidentInBytes.load()), doc.GetAllocator());
+                poolObject.AddMember(FragmentationAttribStr, heapMemoryUsageHost.m_fragmentation, doc.GetAllocator());
+                poolObject.AddMember(UniqueAllocationsInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageHost.m_uniqueAllocationBytes.load()), doc.GetAllocator());
+            }
+            else if (const auto& heapMemoryUsageDevice = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device); heapMemoryUsageDevice.m_totalResidentInBytes > 0)
+            {
+                poolObject.AddMember(MemoryTypeAttribStr,               DeviceMemoryTypeValueStr, doc.GetAllocator());
```
Comment (gadams3) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp:65]:
```suggestion
                poolObject.AddMember(MemoryTypeAttribStr, DeviceMemoryTypeValueStr, doc.GetAllocator());
```

**Review by gadams3 — APPROVED** [review]

File: Gems/Atom/Utils/Code/Source/ImGuiGpuProfiler.cpp [hunk 7]
```diff
@@ -1769,64 +1743,8 @@ namespace AZ
...
-                    imagesArray.PushBack(imageObject, doc.GetAllocator());
-                }
-                poolObject.AddMember(ImagesListAttribStr, imagesArray, doc.GetAllocator());
-                poolsArray.PushBack(poolObject, doc.GetAllocator());
-            }
-            root.AddMember(PoolsAttribStr, poolsArray, doc.GetAllocator());
-            root.AddMember(MemoryDataVersionMajorAttribStr, 1, doc.GetAllocator());
-            root.AddMember(MemoryDataVersionMinorAttribStr, 0, doc.GetAllocator());
-            root.AddMember(MemoryDataVersionRevisionAttribStr, 0, doc.GetAllocator());
+            AZ::RHI::WritePoolsToJson(m_savedPools, doc, root);
```
Comment (moudgils) [inline Gems/Atom/Utils/Code/Source/ImGuiGpuProfiler.cpp:1747]: Can you use RHISysteminterface api here?

**Review by moudgils — APPROVED** [review]
Looks good. Just a small comment about using RHISysteminterface to do the dump from imguigpuprofiler

**Review by moudgils — APPROVED** [review]
Nice. Approved

**Pull request merged**